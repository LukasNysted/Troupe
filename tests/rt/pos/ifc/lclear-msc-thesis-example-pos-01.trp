(* 

Created: 2021-08-06; updated 2021-08-09; AA

This test case demonstrates the need to add lclear label to the 
values obtained through the mailbox (which includes both `peek`
and `consume` operations)

The example is based on the masters thesis by Martin, Kristian, and Emil
on the formalization of Troupe in Coq. 
   
*)

(* I will need two processes *)

let fun f () = 
	let  val c = raisembox (`{bob}`)
	     val x = consume  (0, `{alice}`, `{alice,bob}`) 
	     val _ = debugpc ()
	     val _ = blockdeclto (authority , `{alice}`)
	     val _ = lowermbox (c, authority)
	     val _ = debugpc ()
	     val _ = peek  (0, `{alice}`, `{alice}`)
	     val _ = debugpc ()
	in ladv (`{alice}`, ()) 
	end 
    val p1 = spawn f 
    fun g () = if true raisedTo `{alice}` then 
    			send (p1, "HELLO");
               		send (p1, "BYE")
	        else () 
    val p2 = spawn g 
in (p1, p2)
end
