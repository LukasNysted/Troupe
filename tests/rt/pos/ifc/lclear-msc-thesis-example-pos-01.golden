2021-08-09T20:50:39.717Z [RTM] [32minfo[39m: Skipping network creation. Observe that all external IO operations will yield a runtime error.
>>> Main thread finished with value: (b39d3b56-1f35-41d5-b78a-fd78e22d0a44@{}%{}, 61c3f51e-51ec-407c-b398-b61c1fe94efc@{}%{})@{}%{}
[31mPID:b39d3b56-1f35-41d5-b78a-fd78e22d0a44@{}%{}    [39m [31mPC:{}               [39m [31mBL:{alice,bob}      [39m [31mHNNORMAL            [39m [31m_sp:21              [39m 
[31mPID:b39d3b56-1f35-41d5-b78a-fd78e22d0a44@{}%{}    [39m [31mPC:{}               [39m [31mBL:{alice}          [39m [31mHNNORMAL            [39m [31m_sp:21              [39m 
[31mPID:b39d3b56-1f35-41d5-b78a-fd78e22d0a44@{}%{}    [39m [31mPC:{}               [39m [31mBL:{alice}          [39m [31mHNNORMAL            [39m [31m_sp:21              [39m 
